### PR TITLE
improve AutoRelay configuration, add config option for static relays

### PR DESCRIPTION
**Pull request by marten-seemann**

### CODE REVIEW

**Review by vyzo — APPROVED** [review]

File: swarm.go [hunk 1]
```diff
@@ -22,9 +22,16 @@ type SwarmConfig struct {
...
 	// it's not reachable from the public internet.
+	//
+	// Deprecated: This flag is deprecated and is overriden by
+	// `Swarm.AutoRelay.Enabled` if specified.
 	EnableAutoRelay bool
 
-	// RelayService.* controls the "auto relay service" feature.
+	// AutoRelay controls the "auto relay service" feature.
+	// When enabled, the node will use relays if it is not publicly reachable.
+	AutoRelay AutoRelay
```
Comment (lidel) [inline swarm.go:32]: I am worried about ambiguity, better to be very clear around relay "service" and relay "client". 
 Would it be ok if we rename `AutoRelay` to more explicit `RelayClient`? 
 
```suggestion
	RelayClient RelayClient
```

This way, we will have `RelaySevice` and `RelayClient` sections that are easier to wrap the head around without reading docs.

File: swarm.go [hunk 2]
```diff
@@ -22,9 +22,16 @@ type SwarmConfig struct {
 	// EnableAutoRelay enables the "auto relay user" feature.
 	// Node will find and use advertised public relays when it determines that
 	// it's not reachable from the public internet.
+	//
+	// Deprecated: This flag is deprecated and is overriden by
+	// `Swarm.AutoRelay.Enabled` if specified.
 	EnableAutoRelay bool
```
Comment (lidel) [inline swarm.go:28]: nit: we should do what we already did for `DisableRelay` and hide it from new configs:
```suggestion
	EnableAutoRelay bool `json:",omitempty"`
```

**Review by lidel — CHANGES_REQUESTED** [review]

File: swarm.go (hunk 1 above)
Comment (marten-seemann) [reply swarm.go:32]: I'm ok with this. @vyzo, wdyt?

Comment (vyzo) [reply swarm.go:32]: yeah, fine by me.

Comment (lidel) [reply swarm.go]: Applied  change in https://github.com/ipfs/go-ipfs-config/pull/154/commits/11644afd3c4e800716e8260357d5caa48fdfe90f

**Comment by aschmahmann** [comment]
@marten-seemann is this ready to merge, or do you plan on adding any other config related to the go-libp2p upgrade?

@lidel is fine with how the config is setup and we'd like to do a release so we can unblock other things that rely on the various go-ipfs-config changes.

If you have more changes to make that's fine, we'll have to partially revert #146 (i.e. the libp2p things, but not the OptionalInteger changes), release and then re-release later.

**Comment by marten-seemann** [comment]
> @marten-seemann is this ready to merge, or do you plan on adding any other config related to the go-libp2p upgrade?

This is ready to merge. I'm planning to add another option for hole punching tomorrow, but that's a different thing, so it should be a separate PR.

> If you have more changes to make that's fine, we'll have to partially revert #146 (i.e. the libp2p things, but not the OptionalInteger changes), release and then re-release later.

Not sure I understand. The changes in #146 are all correct, as far as I can tell, and needed by the go-libp2p update.

**Pull request merged**